protocols/dcutr/src/behaviour.rs

```diff
@@ -0,0 +1,401 @@
+        remote_peer_id: PeerId,
+        error: UpgradeError,
+    },
+}
+
+#[derive(Debug, Error)]
+pub enum UpgradeError {
+    #[error("Failed to dial peer.")]
+    Dial,
+    #[error("Failed to establish substream: {0}.")]
+    Handler(ProtocolsHandlerUpgrErr<void::Void>),
+}
+
+pub struct Behaviour {
+    /// Queue of actions to return when polled.
+    queued_actions: VecDeque<Action>,
+
+    /// All direct (non-relayed) connections.
+    direct_connections: HashMap<PeerId, HashSet<ConnectionId>>,
+}
+
+impl Behaviour {
+    pub fn new() -> Self {
+        Behaviour {
+            queued_actions: Default::default(),
+            direct_connections: Default::default(),
+        }
+    }
+}
+
+impl NetworkBehaviour for Behaviour {
+    type ProtocolsHandler = handler::Prototype;
+    type OutEvent = Event;
+
+    fn new_handler(&mut self) -> Self::ProtocolsHandler {
+        handler::Prototype::UnknownConnection
+    }
+
+    fn addresses_of_peer(&mut self, _peer_id: &PeerId) -> Vec<Multiaddr> {
+        vec![]
+    }
+
+    fn inject_connected(&mut self, _peer_id: &PeerId) {}
+
+    fn inject_connection_established(
+        &mut self,
+        peer_id: &PeerId,
+        connection_id: &ConnectionId,
+        connected_point: &ConnectedPoint,
+        _failed_addresses: Option<&Vec<Multiaddr>>,
+    ) {
+        if connected_point.is_relayed() {
+            if connected_point.is_listener() && !self.direct_connections.contains_key(peer_id) {
+                // TODO: Try dialing the remote peer directly. Specification:
+                //
+                // > The protocol starts with the completion of a relay connection from A to B. Upon
+                // observing the new connection, the inbound peer (here B) checks the addresses
+                // advertised by A via identify. If that set includes public addresses, then A may
+                // be reachable by a direct connection, in which case B attempts a unilateral
+                // connection upgrade by initiating a direct connection to A.
+                //
+                // https://github.com/libp2p/specs/blob/master/relay/DCUtR.md#the-protocol
+                self.queued_actions.push_back(Action::Connect {
+                    peer_id: *peer_id,
+                    attempt: 1,
+                    handler: NotifyHandler::One(*connection_id),
+                });
+                let local_addr = match connected_point {
+                    ConnectedPoint::Listener { local_addr, .. } => local_addr,
+                    ConnectedPoint::Dialer { .. } => unreachable!("Due to outer if."),
+                };
+                self.queued_actions.push_back(
+                    NetworkBehaviourAction::GenerateEvent(Event::InitiateDirectConnectionUpgrade {
+                        remote_peer_id: *peer_id,
+                        local_relayed_addr: local_addr.clone(),
+                    })
+                    .into(),
+                );
```

Comment:
This could be written as a single `.extend` of an array to make it clear that the actions belong together.

Reply:
Sounds good to me. Done across the whole file with b50708b7fe237e59ae549499a6b992a664386b60.